added support for single key strafing

ohmailon opened the pull request:
In Keys2XInput we got the ability to set the options in a way that allows us to strafe while only pressing A or D without the need of pressing W. This should be the same feature but in the wooting style with a fancy slider & a simple option to disable it :)

PS: I've split the AdvancedConfig File into multiple Component files. Makes it a lot easier to navigate between components.

~~ℹ This feature is based on the branch from https://github.com/WootingKb/wooting-double-movement/pull/17 so this https://github.com/WootingKb/wooting-double-movement/pull/17 needs to be handled first.~~

![image](https://user-images.githubusercontent.com/57702380/116479946-c471ba80-a880-11eb-833f-324534d9a696.png)

Commit:
improved application structure & tabs menu

Commit:
implemented single key strafing & made some clean ups

Commit:
added info tooltips

Commit:
merged #17 into #18

Commit:
Merge branch 'feature/binds-per-direction' into feature/advanced-strafe-angle

# Conflicts:
#	src/native/src/controller.rs
#	src/native/src/service.rs
#	src/view/AdvancedSettings.tsx

ohmailon commented:
I merged the changes from #17 into this branch

Commit:
fixed error in tooltip

Commit:
change back to non rawinput in repo

Commit:
fixed wrong default value for normal strafing angle

Commit:
fixed wrong default value for normal strafing angle

Commit:
Merge branch 'feature/binds-per-direction' into feature/advanced-strafe-angle

# Conflicts:
#	src/view/AdvancedSettings.tsx

Commit:
Merge branch 'feature/binds-per-direction' into feature/advanced-strafe-angle

Commit:
prettier cleanup

Commit:
Merge branch 'feature/binds-per-direction' into feature/advanced-strafe-angle

# Conflicts:
#	src/view/AdvancedSettings.tsx

Commit:
merged binds-per-direction

Commit:
Merge branch 'main' into feature/advanced-strafe-angle

Commit:
Merge branch 'feature/binds-per-direction' into feature/advanced-strafe-angle

# Conflicts:
#	src/view/AdvancedSettings.tsx

simon-wh commented:
Could you rebase this on the latest `main` to try clean up the git history a bit?

Commit:
Merge pull request #2 from WootingKb/main

update fork

Commit:
Merge branch 'main' into feature/advanced-strafe-angle

# Conflicts:
#	src/main.ts
#	src/view/components/settings/key-bind/KeyBindControl.tsx

ohmailon commented:
Merged the current main branch into this PR.

KMontag42 commented:
when could this get merged? this is the biggest missing feature from wooting double movement

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+                                This option allows you to adjust the angle you
+                                will strafe by pressing <Kbd>Left</Kbd>/
+                                <Kbd>Right</Kbd> at the same time as{" "}
+                                <Kbd>Forward</Kbd>/<Kbd>Back</Kbd> (e.g.{" "}
```
This should only mention `Forward` as the strafe angle only applies when holding forward. Other than that, looks very good 👍

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+                fontSize="sm"
+                onClick={setDefaultSettings}
+              >
+                Reset settings to Wooting recommended
```
We should make it so that this Reset button only affects the settings on this tab. Currently, if you press reset on either tab, it'll reset everything

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+                              />
+                            </PopoverTrigger>
+                            <PopoverContent backgroundColor={bg}>
+                              <PopoverArrow />
```
The colour of the arrow does not match the background of the Popover when in Dark Mode
![image](https://user-images.githubusercontent.com/10797576/124494452-c2f7cd80-ddae-11eb-8427-5a73a5afe854.png)

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+                          <Text variant="heading">
+                            Enable Single Key Strafing
+                          </Text>
+                          <Popover trigger="hover" placement="top">
```
Any particular reason for using Popover straight as opposed to [Tooltip](https://chakra-ui.com/docs/overlay/tooltip). If there's an intentional reason then I'd suggest making a component for these InfoIcon tooltips so it can be easily reused

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+
+export function AdvancedSettingsCard() {
+  function updateWindowSize(index: ExpandedIndex) {
+    console.log("Accordian index: ", index);
```
Should make this a debug log or remove

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+import { Key } from "ts-keycode-enum";
+import { InfoOutlineIcon } from "@chakra-ui/icons";
+
+function calcDisplay(keyOne: any, def: any) {
```
It would be good to make this strongly typed rather than using `any`

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+    "isAdvancedStrafeOn",
+    false
+  );
+  const [keyMapping, setKeymapping] = keyMappingRemoteValue;
```
Why is this done on a separate line rather than done straight from the `useRemoteValue` call

simon-wh replied:
nvm, I see why now, the value/set is passed on to separate components. It'd be good if the lines were kept together (i.e. the `useRemoteValue` line and the deconstructor line)

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+                    <AngleControl
+                      remoteValue={leftJoystickValues}
+                      min={15}
+                      max={72}
```
It appears that the slider range mentioned here doesn't match the updates ones in the current release https://github.com/WootingKb/wooting-double-movement/blob/main/src/view/AdvancedSettings.tsx#L44

simon-wh commented on src/view/components/AdvancedSettings.tsx:
```diff
@@ -0,0 +1,226 @@
+  const minTabHeight = "180px";
+
+  return (
+    <Card p="2">
```
The HTML part of this component is quite big, it'd be good if it was split up into smaller more processeable chunks

simon-wh replied:
e.g. seperate components for the main layout, each tab, the main strafe angle setting and the single key strafing setting

simon-wh commented on src/native/src/service.rs:
```diff
@@ -168,14 +169,14 @@ impl Service {
 
             #[cfg(not(feature = "ds4"))]
-            let report = self
+                let report = self
```
What's up with this indenting?

simon-wh commented on src/native/src/controller.rs:
```diff
@@ -178,10 +178,14 @@ impl JoystickState {
     }
 
-    pub fn get_basic_direction(&self, config: Option<&JoystickAngleConfiguration>) -> (f32, f32) {
+    pub fn get_basic_direction(&self, config: Option<&ServiceConfiguration>) -> (f32, f32) {
```
It feels a little weird that the entire ServiceConfiguration is getting passed through here. It feels like only the subsection of the configuration that is relevant for the direction calculation should be passed. That can be achieved through restructuring the `ServiceConfiguration` to have a sub-struct that contains all the relevant parts for this, although that has an annoying rippling affect throughout all these changes. Alternatively it could get passed the sub-parts as different parameters to avoid the configuration structure change.

This is up for discussion so don't feel like you need to do a massive refactor for this

simon-wh commented on src/view/components/settings/angle/AngleSlider.tsx:
```diff
@@ -0,0 +1,73 @@
+      <NumberInput
+        onChange={(_, value) => {
+          if (!Number.isNaN(value)) {
+            console.log(value);
```
We should either remove this or change it to debug log

simon-wh reviewed (changes requested):
Overall looks really good, mainly have minor code things that need changing. Still need to confirm the UI styling/layout to see if there are anymore tweaks needed on that front

ohmailon committed:
- adding per settings reset function
- fixed text in ? hint

simon-wh committed:
Fix small compilation issues

simon-wh committed:
Fix tooltip arrow colour being wrong in dark mode

simon-wh committed:
Wee log cleanup / formatting

simon-wh committed:
Adjust main strafe angle range to match current main

simon-wh committed:
Split up AdvancedSettings into smaller components and cleanup structure

- Also fix listener for setting value changes not triggering if the window changes the value

simon-wh committed:
Move popover tooltip to InfoTooltip component, use accent color

simon-wh committed:
calcDisplay -> keybindDisplay and make it strongly typed

simon-wh committed:
Cleanup structure of config & fix angle slider value update spam

simon-wh committed:
Cleanup logic for calculating advanced strafe angle

simon-wh reviewed (approved):
Did some of the change I requested to get this wrapped up. Thanks again for the contribution!
